File: src/cn/leetcode/codes/simple7/Simple7.java
package cn.leetcode.codes.simple7;

public class Simple7 {


    public static void main(String[] args) {


        System.out.println(reverse(123));
        System.out.println(reverse(-123));
        System.out.println(reverse(130));

        System.out.println(" JDK ---");

//        System.out.println(Integer.reverse(123));
//        System.out.println(Integer.reverse(-123));
//        System.out.println(Integer.reverse(130));

    }


    public static int reverse(int x) {
        int rev = 0;
        while (x != 0) {
            // 取出个位数数据
            int pop = x % 10;

            //不断地取出个位数
            x /= 10;

            if (rev > Integer.MAX_VALUE / 10 || (rev == Integer.MAX_VALUE / 10 && pop > 7)){
                return 0;
            }
            if (rev < Integer.MIN_VALUE / 10 || (rev == Integer.MIN_VALUE / 10 && pop < -8)){
                return 0;
            }
            rev = rev * 10 + pop;
        }
        return rev;
    }

    /*
    作者：LeetCode
    链接：https://leetcode-cn.com/problems/reverse-integer/solution/zheng-shu-fan-zhuan-by-leetcode/
    来源：力扣（LeetCode）
    著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
    */
}

File: src/cn/leetcode/codes/simple7/SImple7.kt
package cn.leetcode.codes.simple7

import cn.leetcode.codes.out

/*

7. 整数反转
给出一个 32 位的有符号整数，你需要将这个整数中每位上的数字进行反转。

示例 1:

输入: 123
输出: 321
 示例 2:

输入: -123
输出: -321
示例 3:

输入: 120
输出: 21
注意:

假设我们的环境只能存储得下 32 位的有符号整数，则其数值范围为 [−231,  231 − 1]。请根据这个假设，如果反转后整数溢出那么就返回 0。


 */

fun main(){

    out(reverse(123))


}


fun reverse(x: Int): Int {
    var ans = 0

    var y = x
    while (y != 0){
        val pop = x % 10
        //溢出检查
        if (ans > Int.MAX_VALUE / 10 || (ans == Int.MAX_VALUE / 10 && pop > 7)){
            return 0
        }
        if (ans < Int.MIN_VALUE / 10 || (ans == Int.MIN_VALUE / 10 && pop < -8)){
            return 0
        }
        ans = ans * 10 + pop
        y /= 10
    }
    return ans
}